tests/Unit/DataStructures/Tensor/Expressions/Test_Contract.cpp

```diff
@@ -0,0 +1,540 @@
+    }
+  }
+
+  // Contract second and fourth indices of (lower, upper, lower, lower) tensor
+  // to rank 2 tensor
+  Tensor<DataType, Symmetry<4, 3, 2, 1>,
+         index_list<SpatialIndex<3, UpLo::Lo, Frame::Inertial>,
+                    SpatialIndex<3, UpLo::Up, Frame::Inertial>,
+                    SpatialIndex<4, UpLo::Lo, Frame::Inertial>,
+                    SpatialIndex<3, UpLo::Lo, Frame::Inertial>>>
+      Rlull(used_for_size);
+  create_tensor(make_not_null(&Rlull));
+
+  const Tensor<DataType, Symmetry<2, 1>,
+               index_list<SpatialIndex<3, UpLo::Lo, Frame::Inertial>,
+                          SpatialIndex<4, UpLo::Lo, Frame::Inertial>>>
+      RkJij_contracted = TensorExpressions::evaluate<ti_k_t, ti_i_t>(
+          Rlull(ti_k, ti_J, ti_i, ti_j));
+
+  for (size_t k = 0; k < 3; k++) {
+    for (size_t i = 0; i < 4; i++) {
+      DataType expected_sum = make_with_value<DataType>(used_for_size, 0.0);
+      for (size_t j = 0; j < 3; j++) {
+        expected_sum += Rlull.get(k, j, i, j);
+      }
+      CHECK(RkJij_contracted.get(k, i) == expected_sum);
+    }
+  }
+
+  // Contract third and fourth indices of (upper, lower, lower, upper) tensor to
+  // rank 2 tensor
+  Tensor<DataType, Symmetry<3, 2, 2, 1>,
+         index_list<SpacetimeIndex<4, UpLo::Up, Frame::Inertial>,
+                    SpacetimeIndex<3, UpLo::Lo, Frame::Inertial>,
+                    SpacetimeIndex<3, UpLo::Lo, Frame::Inertial>,
+                    SpacetimeIndex<3, UpLo::Up, Frame::Inertial>>>
+      Rullu(used_for_size);
+  create_tensor(make_not_null(&Rullu));
+
+  const Tensor<DataType, Symmetry<2, 1>,
+               index_list<SpacetimeIndex<4, UpLo::Up, Frame::Inertial>,
+                          SpacetimeIndex<3, UpLo::Lo, Frame::Inertial>>>
+      RFcgG_contracted = TensorExpressions::evaluate<ti_F_t, ti_c_t>(
+          Rullu(ti_F, ti_c, ti_g, ti_G));
+
+  for (size_t f = 0; f < 5; f++) {
+    for (size_t c = 0; c < 4; c++) {
+      DataType expected_sum = make_with_value<DataType>(used_for_size, 0.0);
+      for (size_t g = 0; g < 4; g++) {
+        expected_sum += Rullu.get(f, c, g, g);
+      }
+      CHECK(RFcgG_contracted.get(f, c) == expected_sum);
+    }
+  }
+
+  // Contract first and second indices of (upper, lower, upper, upper) tensor to
+  // rank 2 tensor and reorder indices
+  Tensor<DataType, Symmetry<3, 2, 3, 1>,
+         index_list<SpatialIndex<3, UpLo::Up, Frame::Grid>,
+                    SpatialIndex<3, UpLo::Lo, Frame::Grid>,
+                    SpatialIndex<3, UpLo::Up, Frame::Grid>,
+                    SpatialIndex<2, UpLo::Up, Frame::Grid>>>
+      Ruluu(used_for_size);
+  create_tensor(make_not_null(&Ruluu));
+
+  const Tensor<DataType, Symmetry<2, 1>,
+               index_list<SpatialIndex<2, UpLo::Up, Frame::Grid>,
+                          SpatialIndex<3, UpLo::Up, Frame::Grid>>>
+      RKkIJ_contracted_to_IJ = TensorExpressions::evaluate<ti_J_t, ti_I_t>(
```

Comment:
-> "RKkIJ_contracted_to_JI" ? (At least that's what I expected this would be called, and how you've done it in the cases below.)

Reply:
Done!